### PR TITLE
feat(store-sync): wait for idle after each chunk of logs in a block

### PR DESCRIPTION
**Pull request by Kooshaba**
don't ask me why or how, but adding `sleep(1)` after calling the `storageAdapter` made hydration updates show up in the client. i was going crazy trying to figure out why i wasn't receiving them in the main thread, but then remembered having a similar problem in DF with how async calls were being handled. if someone actually knows the mechanism that is being abused here lmk and we can come up with a less obscure solution.

also once i started receiving updates i realized the % calculation was wrong, so fixed that too.

getting these updates is important so i can properly benchmark how long hydration takes on the sky strife end. my manual calculations on my pc are ~5 seconds.

### CODE REVIEW

File: packages/store-sync/src/createStoreSync.ts [hunk 1]
```diff
@@ -145,11 +149,12 @@ export async function createStoreSync<TConfig extends StoreConfig = StoreConfig>
         await storageAdapter({ blockNumber, logs: chunk });
         onProgress?.({
           step: SyncStep.SNAPSHOT,
-          percentage: ((i + chunk.length) / chunks.length) * 100,
+          percentage: Math.floor((i / chunks.length) * 100),
           latestBlockNumber: 0n,
           lastBlockNumberProcessed: blockNumber,
           message: "Hydrating from snapshot",
         });
+        await sleep(1);
```
Comment (holic) [inline packages/store-sync/src/createStoreSync.ts:157]: Can we use waitForIdle here from common package instead?

File: packages/store-sync/src/createStoreSync.ts [hunk 2]
```diff
@@ -145,11 +149,12 @@ export async function createStoreSync<TConfig extends StoreConfig = StoreConfig>
         await storageAdapter({ blockNumber, logs: chunk });
         onProgress?.({
           step: SyncStep.SNAPSHOT,
-          percentage: ((i + chunk.length) / chunks.length) * 100,
+          percentage: Math.floor((i / chunks.length) * 100),
```
Comment (holic) [inline packages/store-sync/src/createStoreSync.ts:152]: I am intentionally not rounding here so frontends can decide how much precision they want to render. Let’s remove this diff.

File: packages/store-sync/src/createStoreSync.ts (hunk 1 above)
Comment (holic) [reply packages/store-sync/src/createStoreSync.ts:157]: And can we also add a note here around why we’re doing this? Something like

> RECS is a synchronous API so hydrating in a loop like this blocks downstream render cycles that would display the percentage climbing up to 100. We wait for idle callback here to give rendering a chance to complete.

**Comment by holic** [comment]
Wonder if we should wait on every block or some number of iterations. Just thinking about how hydrating thousands of blocks will take significantly longer to complete if we have to wait for idle on each block.

File: packages/store-sync/src/createStoreSync.ts [hunk 3]
```diff
@@ -145,11 +145,16 @@ export async function createStoreSync<TConfig extends StoreConfig = StoreConfig>
         await storageAdapter({ blockNumber, logs: chunk });
         onProgress?.({
           step: SyncStep.SNAPSHOT,
-          percentage: ((i + chunk.length) / chunks.length) * 100,
+          percentage: (i / chunks.length) * 100,
```
Comment (holic) [inline packages/store-sync/src/createStoreSync.ts:148]:
```suggestion
          percentage: ((i  + 1) / chunks.length) * 100,
```

File: packages/store-sync/src/createStoreSync.ts [hunk 4]
```diff
@@ -145,11 +145,16 @@ export async function createStoreSync<TConfig extends StoreConfig = StoreConfig>
         await storageAdapter({ blockNumber, logs: chunk });
         onProgress?.({
           step: SyncStep.SNAPSHOT,
-          percentage: ((i + chunk.length) / chunks.length) * 100,
+          percentage: ((i  + 1) / chunks.length) * 100,
```
Comment (holic) [inline packages/store-sync/src/createStoreSync.ts:148]:
```suggestion
          percentage: ((i + 1) / chunks.length) * 100,
```

**Comment by holic** [comment]
waiting on https://github.com/latticexyz/mud/pull/2264 just so we can have a test that this does what we expect

**Review by yonadaaa — APPROVED** [review]
I added a spec and tested manually with Sky Strife, works great 👍 


https://github.com/latticexyz/mud/assets/29184158/400d3dff-758b-47e2-a9ce-49d694a776c7

**Review by holic — APPROVED** [review]

**Pull request merged**